### PR TITLE
"Reserve" the method store when constructing methods

### LINKED ISSUE #18152: Core fetching synchronisation still has a race condition
[ This is a follow-up on #18150 ]

In the core method construction loop, we're testing and setting so called operation bits to avoid having methods recreated unnecessarily.  Setting and testing these bits themselves is guarded by a lock, but further than that, there is no synchronisation.

In short, what happens is the following:

1.  Check the appropriate operation bit for the methods about the be constructed
2.  If the operation bit was unset, construct methods
3.  Set the operation bit

With multiple threads doing this simultaneously, it should be obvious that several might end up seeing the operation bit cleared, and thereby constructing the same methods concurrently.

An idea to fix this is to have a construction lock, making the process this:

0.  Lock the construction lock
1.  Check the appropriate operation bit for the methods about the be constructed
2.  If the operation bit was unset, construct methods
3.  Set the operation bit
4.  Unlock the construction lock

#18151 refactors `algorithm_do_this()` in a way that may make it easy enough to add this sort of lock.

Ref: https://github.com/openssl/openssl/blob/c29cf39449f78008e39af8f83760f2464815248b/crypto/core_algorithm.c#L45-L78

### PR DESCRIPTION
**Pull request by levitte**
Introducing the concept of reserving the store where a number of
provided operation methods are to be stored.

This avoids racing when constructing provided methods, which is
especially pertinent when multiple threads are trying to fetch the
same method, or even any implementation for the same given operation
type.

This introduces a |biglock| in OSSL_METHOD_STORE, which is separate
from the |lock| which is used for more internal and finer grained
locking.

Fixes #18152

### CODE REVIEW

**Comment by levitte** [comment]
Note that this is made on top of #18151.  Please review that one first...  this PR will stop being a draft after #18151 has been merged

**Comment by levitte** [comment]
> Given the amount of discussion we had to come up with this, this is probably worth a comment.

Sure.  I'll wait 'til #18150 is done, though

**Comment by t8m** [comment]
Having two locks on the same structure can be a source of deadlocks. Are we sure they cannot happen? It is unfortunately not obvious from the PR to me.

**Comment by levitte** [comment]
> Having two locks on the same structure can be a source of deadlocks. Are we sure they cannot happen? It is unfortunately not obvious from the PR to me.

I tried to be careful not to cross-lock them.  If that still happens, then I assume that we would have a deadlock because recursion with the older lock.

File: crypto/property/property.c [hunk 1]
```diff
@@ -64,6 +64,7 @@ struct ossl_method_store_st {
     OSSL_LIB_CTX *ctx;
     SPARSE_ARRAY_OF(ALGORITHM) *algs;
     CRYPTO_RWLOCK *lock;
+    CRYPTO_RWLOCK *biglock;
```
Comment (t8m) [inline crypto/property/property.c:67]: Could you please add some comment(s) here as for what is the purpose of each of the locks?

Comment (kaduk) [reply crypto/property/property.c:67]: Yes, IMO comments about what the lock protects should be required.
Also document any lock ordering considerations with other locks (e.g., when both are locked, biglock must be locked first and unlocked last) and scenarios where there are fields that can be read with either lock held but both must be held for writing.
I think https://github.com/freebsd/freebsd-src/blob/c4a6d412c9c75fb50603f1ae52d6bd78f8510e6c/sys/sys/vnode.h#L78-L176 is a great example of clear locking documentation, and https://github.com/torvalds/linux/blob/feb9c5e19e913b53cb536a7aa7c9f20107bb51ec/Documentation/filesystems/directory-locking.rst covers some situations where interactions between multiple locks must be considered.

I also am a strong proponent for adding as many assertions about lock status as you can (even if compile-time conditional under something other than `NDEBUG`).

Comment (mattcaswell) [reply crypto/property/property.c:67]: Also see below in the OpenSSL code where we have documented the locking requirements:

https://github.com/openssl/openssl/blob/cb2764f2a8165421dc5ab52159af99cbf766fa2c/crypto/provider_core.c#L56-L116

Comment (levitte) [reply crypto/property/property.c:67]: I will get to it, thanks for pointing this out

**Comment by t8m** [comment]
@levitte please address the comments and rebase.

**Comment by levitte** [comment]
Ping

**Review by hlandau — APPROVED** [review]

**Review by t8m — APPROVED** [review]

**Comment by hlandau** [comment]
Merged to master and 3.0. Thank you.

**Comment by levitte** [comment]
Interesting...  one of the pushes has the marker "Verified", which is about signature.  That's the cherry-pick to 3.0, and indeed, `git verify-commit ddb81a94ac8af145750940c20999ac952165bf10` tells me that you indeed signed it, @hlandau.

foo

**Pull request closed without merging**